### PR TITLE
Model Interface

### PR DESCRIPTION
**Pull request by mazhurin**
Model Interface class. Categorical features for spark.iForest.

### CODE REVIEW

File: src/baskerville/models/anomaly_model.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                 scaler_with_mean=False, scaler_with_std=True):
+        super().__init__()
+        self.prediction_column = prediction_column
+        self.score_column = score_column
+        self.num_trees = num_trees
+        self.max_samples = max_samples
+        self.max_features = max_features
+        self.max_depth = max_depth
+        self.contamination = contamination
+        self.bootstrap = bootstrap
```
Comment (mkaranasou) [inline src/baskerville/models/anomaly_model.py:37]: We could have all this set in base, since it is common in both anomaly models to avoid duplication. Unless you think this is too Isolation Forest specific, so it wouldn't make sense. :)

File: src/baskerville/models/anomaly_model.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        self.approximate_quantile_relative_error = approximate_quantile_relative_error
+        self.seed = seed
+        self.scaler_with_mean = scaler_with_mean
+        self.scaler_with_std = scaler_with_std
+        self.features = features
+        self.categorical_features = categorical_features
+        self.feature_map_column = feature_map_column
+
+        self.storage_level = StorageLevel.OFF_HEAP
```
Comment (mkaranasou) [inline src/baskerville/models/anomaly_model.py]: Shouldn't this be configurable? I see there is a `set_storage_level` method, but why not using the configuration for this? Unless we need this to behave differently than the rest of the app?

File: src/baskerville/models/anomaly_model.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            self.indexes.append(indexer.fit(df))
+
+    def _add_categorical_features(self, df, feature_column):
+        index_columns = []
+        for index_model in self.indexes:
+            df = index_model.transform(df)
+            index_columns.append(index_model.getOutputCol())
+
+        add_categories = F.udf(lambda features, arr: Vectors.dense(np.append(features, [v for v in arr])),
+                               VectorUDT())
```
Comment (mkaranasou) [inline src/baskerville/models/anomaly_model.py]: 👍 
Do you think this could be moved under `spark/udfs`, or is it one of those cases where we need to have it here for it to work properly?

File: src/baskerville/models/anomaly_model.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        df = df.withColumn('features_all', add_categories(feature_column, array(*index_columns))) \
+            .drop(*index_columns) \
+            .drop(feature_column) \
```
Comment (mkaranasou) [inline src/baskerville/models/anomaly_model.py]: It *might* be worth it (or not) to have this in one line
```python
.drop(*index_columns) \
.drop(feature_column)
```
```python 
.drop(*index_columns, feature_column) \
```
but I am not 100% sure if spark is spark enough to concatenate the two drops or if it matters or in general.

Anyway, pretty minor, just adding this comment here to keep in mind.

File: src/baskerville/models/anomaly_model.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            maxDepth=self.max_depth,
+            contamination=self.contamination,
+            bootstrap=self.bootstrap,
+            approxQuantileRelativeError=self.approximate_quantile_relative_error,
+            numCategoricalFeatures=len(self.categorical_features)
+        )
+        iforest.setSeed(self.seed)
+        params = {'threshold': self.threshold}
+        self.iforest_model = iforest.fit(df, params)
+        df.unpersist()
```
Comment (mkaranasou) [inline src/baskerville/models/anomaly_model.py:119]: Is this the last part where `df` is used? (wondering if it makes sense to unpersist here)

File: src/baskerville/models/anomaly_model.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        self.iforest_model.write().overwrite().save(os.path.join(path, 'iforest'))
+        self.scaler_model.write().overwrite().save(os.path.join(path, 'scaler'))
+
+        if len(self.categorical_features):
+            for feature, index in zip(self.categorical_features, self.indexes):
+                index.write().overwrite().save(os.path.join(path, 'indexes', feature))
+
+    def load(self, path, spark_session=None):
+        self.iforest_model = IForestModel.load(os.path.join(path, 'iforest'))
+        self.scaler_model = StandardScalerModel().load(os.path.join(path, 'scaler'))
```
Comment (mkaranasou) [inline src/baskerville/models/anomaly_model.py]: Very very minor, I think `StandardScalerModel` doesn't need to be instantiated for load?

Comment (mkaranasou) [reply src/baskerville/models/anomaly_model.py]: Also we can avoid all of the `os.path.join(path, 'iforest or scaler')` by using: `get_classifier_load_path(path)` and `get_scaler_load_path(path)` (perhaps the naming is not the best for these functions.. 🤔 but anyway, I just thought we could avoid having the `os.path...` everywhere)

File: src/baskerville/models/anomaly_model.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    def save(self, path, spark_session=None):
+        file_manager = FileManager(path, spark_session)
+        file_manager.save_to_file(self.get_params(), os.path.join(path, 'params.json'), format='json')
+        self.iforest_model.write().overwrite().save(os.path.join(path, 'iforest'))
+        self.scaler_model.write().overwrite().save(os.path.join(path, 'scaler'))
+
+        if len(self.categorical_features):
+            for feature, index in zip(self.categorical_features, self.indexes):
+                index.write().overwrite().save(os.path.join(path, 'indexes', feature))
```
Comment (mkaranasou) [inline src/baskerville/models/anomaly_model.py]: Aren't we saving anything in the `Model` 's table? Or did I miss it somewhere?

File: src/baskerville/models/anomaly_model_sklearn.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        ).withColumn(
+            self.score_column, F.col('y.score')
+        ).drop('y')
+        return df
+
+    def save(self, path, spark_session=None):
+        file_manager = FileManager(path, spark_session)
+        file_manager.save_to_file(self.get_params(), os.path.join(path, 'params.json'), format='json')
+        file_manager.save_to_file(self.iforest_model, os.path.join(path, 'iforest.pickle'), format='pickle')
+        file_manager.save_to_file(self.scaler_model, os.path.join(path, 'scaler.pickle'), format='pickle')
```
Comment (mkaranasou) [inline src/baskerville/models/anomaly_model_sklearn.py:161]: Same thing here, aren't we saving anything in database?

File: src/baskerville/spark/schemas.py [hunk 9]
```diff
@@ -13,23 +13,6 @@
...
-                T.StructField("precision", T.DoubleType(), True),
-                T.StructField("f1_score", T.DoubleType(), True),
-                T.StructField("classifier", T.BinaryType(), True),
-                T.StructField("scaler", T.BinaryType(), True),
-                T.StructField("n_training", T.IntegerType(), True),
-                T.StructField("n_testing", T.IntegerType(), True),
-                T.StructField("threshold", T.DoubleType(), True)
-            ])
-
-
```
Comment (mkaranasou) [inline src/baskerville/spark/schemas.py:32]: Why are we deleting this? Aren't we going to save to `models` table?

File: src/baskerville/util/baskerville_tools.py [hunk 10]
```diff
@@ -90,8 +85,11 @@ def get_request_sets(self):
     def get_latest_ml_model_from_db(self):
         return self.session.query(Model).order_by(Model.id.desc()).first()
 
-    def get_ml_model_from_db(self, version_id):
-        return self.session.query(Model).filter(Model.id == version_id).first()
+    def get_ml_model_from_db(self, model_id):
+        if model_id <= 0:
```
Comment (mkaranasou) [inline src/baskerville/util/baskerville_tools.py]: `if model_id < 0` (we need -1 to bring the latest right? not that `<=` won't work in our case but it might have unintended consequences :) )

File: src/baskerville/util/file_manager.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+import _pickle as cPickle
+
+import errno
+
+
+class FileManager(object):
+
+    def __init__(self, path, spark_session=None):
+        """
+        FileManager supports both local and distributed file systems.
```
Comment (mkaranasou) [inline src/baskerville/util/file_manager.py:14]: Personally, I would prefer to have polymorphism here and avoid all `ifs` in the code but this will do the work well too. :)

File: src/baskerville/util/helpers.py [hunk 12]
```diff
@@ -254,12 +253,7 @@ def get_default_data_path() -> str:
     Returns the absolute path to the data folder
     :return:
     """
-    baskerville_root = os.environ.get('BASKERVILLE_ROOT')
-    if baskerville_root:
-        return os.path.join(baskerville_root, 'data')
```
Comment (mkaranasou) [inline src/baskerville/util/helpers.py]: we need this for docker, because docker moves things in weird folders and some things like the partitioning that's looking for the templates in the data folder will fail.

File: src/baskerville/util/helpers.py [hunk 13]
```diff
@@ -254,12 +253,7 @@ def get_default_data_path() -> str:
...
-    return os.path.join(
-        os.path.dirname(os.path.realpath(__file__)), '..', '..', '..', 'data'
-    )
```
Comment (mkaranasou) [inline src/baskerville/util/helpers.py:262]: This ^ is safer than `f'{os.path.dirname(os.path.realpath(__file__))}/../../../data'` right?

File: src/baskerville/util/model_serialization.py [hunk 14]
```diff
@@ -6,19 +6,16 @@
 from baskerville.db.models import Model
 from baskerville.spark import get_or_create_spark_session
 from baskerville.util.enums import AlgorithmEnum
-from baskerville.util.helpers import get_default_data_path, \
-    get_classifier_load_path, get_scaler_load_path
+from baskerville.util.helpers import get_default_data_path
 from pyspark.ml.feature import StandardScalerModel
 
 
-def pickle_model(model_id, db_config, out_path, ml_model_out_path):
```
Comment (mkaranasou) [inline src/baskerville/util/model_serialization.py:14]: Oh.. okay this whole file changes must be a merge gone wrong?

File: tests/unit/baskerville_tests/models_tests/test_base_spark.py [hunk 15]
```diff
@@ -758,50 +771,6 @@ def test_save(self):
 
         self.spark_pipeline.refresh_cache.assert_called_once()
 
-    @mock.patch('baskerville.models.base_spark.F.udf')
-    def test_predict_no_ml_model(self, mock_udf):
```
Comment (mkaranasou) [inline tests/unit/baskerville_tests/models_tests/test_base_spark.py]: Why is this removed? We do have a case where we can run without an ML model, right?

Comment (mkaranasou) [reply tests/unit/baskerville_tests/models_tests/test_base_spark.py]: Or is just unsalvageable? 😛  :)

File: tests/unit/baskerville_tests/models_tests/test_feature_manager.py [hunk 16]
```diff
@@ -129,44 +123,6 @@ def test_get_active_features_no_ml_model_unknown_features(self):
             4
         )
 
-    def test_get_active_features_all_features(self):
```
Comment (mkaranasou) [inline tests/unit/baskerville_tests/models_tests/test_feature_manager.py:132]: Could you please remind me how this will all work with the model features and the extra features?

File: tests/unit/baskerville_tests/models_tests/test_training_pipelines.py [hunk 17]
```diff
@@ -1,107 +1,60 @@
-import sys
```
Comment (mkaranasou) [inline tests/unit/baskerville_tests/models_tests/test_training_pipelines.py:1]: This file probably is a merge fluke? :)

File: tests/unit/baskerville_tests/utils_tests/test_file_manager.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+        some_dict = {'A': 777}
+        file_name = os.path.join(temp_dir, 'file_manager_test.pickle')
+
+        fm.save_to_file(value=some_dict, file_name=file_name, format='pickle')
+        res = fm.load_from_file(file_name, format='pickle')
+        self.assertDictEqual(res, some_dict)
+
+    # HDFS tests are commented our since it should no be executed with all the unit tests
+    # def test_json_hdfs(self):
```
Comment (mkaranasou) [inline tests/unit/baskerville_tests/utils_tests/test_file_manager.py:38]: We should mock the hdfs stuff - or move these tests under entity / functional tests :)
Cool that you implemented them :) 👍

**Review by mkaranasou — COMMENTED** [review]
Hey, great effort in wrapping the model stuff up together 👏 👍 
Just some comments - I know some of the stuff here we have discussed before but please, bear with me :)

File: src/baskerville/models/anomaly_model.py (hunk 1 above)
Comment (mazhurin) [reply src/baskerville/models/anomaly_model.py:37]: Yes, I would prefer to have the base class as generic as possible for flexibility. For example, I keep thinking about the idea that we might want to train an additional separate classifier model with different parameters trained on the previous attacks. It's very easy to train such a model and it might help us to get less false positives when we start banning every anomaly IP.

File: src/baskerville/models/anomaly_model_sklearn.py (hunk 8 above)
Comment (mazhurin) [reply src/baskerville/models/anomaly_model_sklearn.py:161]: No, we are not saving a model in the database. We save only metadata, or training context if you will (the dataset used for training, the metrics used, the performance, comments). Plus the path to storage where the model itself is saved. Chicken and egg issue, remember? This makes sense since we will be able to directly load a model from storage without accessing the database if you know the path. In fact, I already do this in my notebooks.

File: src/baskerville/spark/schemas.py (hunk 9 above)
Comment (mazhurin) [reply src/baskerville/spark/schemas.py:32]: Yes, we are saving in the database, but now in `TrainingPipeline.save()` with our SQLAlchemy class Model. No need for this anymore.

File: src/baskerville/util/file_manager.py (hunk 11 above)
Comment (mazhurin) [reply src/baskerville/util/file_manager.py:14]: I agree. This file is not good. I don't like it either. In fact, JVM class supports local files itself. So, the proper way is to delegate him local file system as well. And yes, as you can see, they do have a proper factory and polymorphism. :-) It will require some code changes and testing thought.  So, I decided to keep it with these terrible if(s) for now. Otherwise we are going to have our factory on top of JVM factory. :-)

File: src/baskerville/models/anomaly_model.py (hunk 2 above)
Comment (mazhurin) [reply src/baskerville/models/anomaly_model.py]: It should. Done.

File: src/baskerville/models/anomaly_model.py (hunk 3 above)
Comment (mazhurin) [reply src/baskerville/models/anomaly_model.py]: I would rather do the opposite and move `to_dense_vector_udf` from `spark/udfs` to `anomaly_model.py` and delete the import `spark/udfs` here. I feel like these 2 udfs are not going to be used anywhere and solely belong to the model implementation file. But you are right, we need to be consistent one way or another. It's not a big deal, I moved it.

File: src/baskerville/models/anomaly_model.py (hunk 4 above)
Comment (mazhurin) [reply src/baskerville/models/anomaly_model.py]: Good catch. Done!

File: src/baskerville/models/anomaly_model.py (hunk 5 above)
Comment (mazhurin) [reply src/baskerville/models/anomaly_model.py:119]: Right. Should not be here. I call persist both in train/predict in `build_features_vectors`. For train() we don't need it to be persisted anymore. I moved unpersist to the training pipeline. Do you think it's OK?

File: src/baskerville/models/anomaly_model.py (hunk 6 above)
Comment (mazhurin) [reply src/baskerville/models/anomaly_model.py]: Done. +1

File: src/baskerville/util/baskerville_tools.py (hunk 10 above)
Comment (mazhurin) [reply src/baskerville/util/baskerville_tools.py]: Done

File: src/baskerville/util/helpers.py (hunk 12 above)
Comment (mazhurin) [reply src/baskerville/util/helpers.py]: Done

File: src/baskerville/util/helpers.py (hunk 13 above)
Comment (mazhurin) [reply src/baskerville/util/helpers.py:262]: Yes, it's better.

File: src/baskerville/util/model_serialization.py (hunk 14 above)
Comment (mazhurin) [reply src/baskerville/util/model_serialization.py:14]: Yes. We don't need this anymore. The model implementation is responsible for serialization now.
I removed test_model parameter for now. We will need a proper model unit test. For `test_model` parameter I think we first need to implement `model_path` parameter. Then we will be able to load/test/run the pipelines with a testing model saved in the repo.

File: tests/unit/baskerville_tests/models_tests/test_base_spark.py (hunk 15 above)
Comment (mazhurin) [reply tests/unit/baskerville_tests/models_tests/test_base_spark.py]: Yes, I jumped the gun here. I put this test back. BTW, I wanted to talk to you about that threshold. It does not make much sense to store it per row. It will be a parameter in engine config probably. Currently, we don't need it yet. In the dashboard, we set it manually. We will start using it when we implement the challenge logic.

File: tests/unit/baskerville_tests/models_tests/test_feature_manager.py (hunk 16 above)
Comment (mazhurin) [reply tests/unit/baskerville_tests/models_tests/test_feature_manager.py:132]: Yes, I still don't understand what this 'extra' mean.
Here is how I see it. The FeatureManager (running on the client) has a list of features. Ideally, this is a superset of all the features we support. This parameter is `engine.features`.  The training pipeline (running on Bakerstreet) has its own list of features. `training.model_parameter.features`. It might be a subset of all the features if needed. This training features will be owned by the model. It will be saved with the model and used by the model during the prediction time. If at prediction time FeatureManager does not provide some feature(s) the model uses the default values. In this case, we can cherry-pick features for different models, introduce new models with new features, and deploy them without breaking the pipelines.

File: tests/unit/baskerville_tests/models_tests/test_training_pipelines.py (hunk 17 above)
Comment (mazhurin) [reply tests/unit/baskerville_tests/models_tests/test_training_pipelines.py:1]: Yes, I deleted this file.

File: tests/unit/baskerville_tests/utils_tests/test_file_manager.py (hunk 18 above)
Comment (mazhurin) [reply tests/unit/baskerville_tests/utils_tests/test_file_manager.py:38]: Mocking the hdfs is not a real test. It would be nice to have that real functional test. Do you want me to move these tests to the `functional` folder and keep them commented out?

File: tests/unit/baskerville_tests/models_tests/test_base_spark.py (hunk 15 above)
Comment (mkaranasou) [reply tests/unit/baskerville_tests/models_tests/test_base_spark.py]: Sure the threshold is something that's in configuration, it doesn't make sense to have it per row - the dashboard threshold though is something different, right? I mean there is the threshold with which the classifier was trained (e.g. fit parameters) and the threshold we adjust manually for now.

Comment (mazhurin) [reply tests/unit/baskerville_tests/models_tests/test_base_spark.py]: Yes, currently, the dashboard ignores `predictions` and using its own threshold to classify an anomaly from the `score`. So far this is convenient (very easy to manually change directly) since the dashboard itself is sending the notification. But as soon as Baskerville becomes responsible for the challenge/ban commands this threshold has to be in Baskerville configuration. Don't forget, we also have another threshold #2 for attack detection. This threshold #2 defines the maximum portion of anomalies in a batch.

File: tests/unit/baskerville_tests/utils_tests/test_file_manager.py (hunk 18 above)
Comment (mkaranasou) [reply tests/unit/baskerville_tests/utils_tests/test_file_manager.py:38]: Well, that's the purpose of unit tests, right? :) But again, between mocking and doing functional tests, I think the later is preferable. (It will be nice to have more functional tests also)
